include/Gaffer/Spreadsheet.h

```diff
@@ -0,0 +1,219 @@
+//        the distribution.
+//
+//      * Neither the name of John Haddon nor the names of
+//        any other contributors to this software may be used to endorse or
+//        promote products derived from this software without specific prior
+//        written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#ifndef GAFFER_SPREADSHEET_H
+#define GAFFER_SPREADSHEET_H
+
+#include "Gaffer/ComputeNode.h"
+#include "Gaffer/NumericPlug.h"
+#include "Gaffer/TypedObjectPlug.h"
+
+namespace Gaffer
+{
+
+IE_CORE_FORWARDDECLARE( StringPlug )
+
+class GAFFER_API Spreadsheet : public ComputeNode
+{
+
+	public :
+
+		Spreadsheet( const std::string &name=defaultName<Spreadsheet>() );
+		~Spreadsheet() override;
+
+		GAFFER_GRAPHCOMPONENT_DECLARE_TYPE( Gaffer::Spreadsheet, SpreadsheetTypeId, ComputeNode );
+
+		/// Plug types
+		/// ==========
+		///
+		/// The spreadsheet is defined using a hierarchy of specialised plug
+		/// types, organised first by row and then by column.
+
+		class RowPlug;
+
+		/// Top level plug that has a child for each row in the spreadsheet.
+		/// This also provides methods for adding and removing rows and columns.
+		/// Accessed via `Spreadsheet::rowsPlug()`.
+		class RowsPlug : public ValuePlug
+		{
+
+			public :
+
+				RowsPlug( const std::string &name = defaultName<RowsPlug>(), Direction direction = In, unsigned flags = Default );
+
+				GAFFER_PLUG_DECLARE_TYPE( Gaffer::Spreadsheet::RowsPlug, Gaffer::SpreadsheetRowsPlugTypeId, Gaffer::ValuePlug );
+
+				/// Methods for adjusting spreadsheet size
+				/// ======================================
+				///
+				/// Several constraints must be maintained when adjusting the
+				/// size of the spreadsheet, so these dedicated methods should
+				/// be used instead of manual addition of children.
+				///
+				/// These methods are defined on the RowPlug rather than on the
+				/// Spreadsheet so that they can be used for the editing and
+				/// serialisation of promoted plugs.
+
+				size_t addColumn( const ValuePlug *value, IECore::InternedString name = IECore::InternedString() );
+				void removeColumn( size_t columnIndex );
```

Comment:
Just echoing a conversation we had around the mix of index .vs. name for columns.  Would it make sense to move to `removeColumn( name )` and `uniqueColumnName = addColumn( name )`? Then its names throughout? Otherwise, if you know a 'name', you have to go iterate through all the default rows cells to find its index?

Reply:
For the return type of `addColumn()`, I don't think index is any less convenient than name, and it has the added benefit of providing the fastest possible lookups for cells in that column. For `removeColumn()` I can see that it could be convenient to pass a name; indeed, in the single existing call outside of the unit tests, it would have been more convenient. Perhaps `addColumn()` should remain as-is, and we could add an overload for `removeColumn()` that takes a name? Alternatively, we could deal in CellPlugs instead, always from the default row?